pkg/gds/admin_test.go

```diff
@@ -0,0 +1,939 @@
+func (s *gdsTestSuite) doRequest(request *apiRequest, reply interface{}) (res *http.Response) {
+	var body io.ReadWriter
+	var err error
+	require := s.Require()
+
+	// Encode the JSON request
+	if request.in != nil {
+		body = &bytes.Buffer{}
+		err = json.NewEncoder(body).Encode(request.in)
+		require.NoError(err)
+	} else {
+		body = nil
+	}
+
+	// Construct the HTTP request
+	w := httptest.NewRecorder()
+	c, _ := gin.CreateTestContext(w)
+	if request.claims != nil {
+		c.Set(admin.UserClaims, request.claims)
+	}
+	c.Request = httptest.NewRequest(request.method, request.path, body)
+	c.Request.Header.Add("Content-Type", "application/json")
+	if request.headers != nil {
+		for k, v := range request.headers {
+			c.Request.Header.Add(k, v)
+		}
+	}
+	if request.params != nil {
+		for k, v := range request.params {
+			c.Params = append(c.Params, gin.Param{
+				Key:   k,
+				Value: v,
+			})
+		}
+	}
+
+	// Call the admin function and return the response
+	request.fn(c)
+	res = w.Result()
+	defer res.Body.Close()
+	if reply != nil {
+		bytes, err := ioutil.ReadAll(res.Body)
+		require.NoError(err)
+		err = json.Unmarshal(bytes, reply)
+		require.NoError(err)
+	}
+	return res
+}
+
+// Test that we get a good response from ProtectAuthenticate.
+func (s *gdsTestSuite) TestProtectAuthenticate() {
+	require := s.Require()
+	a := initAdmin(s)
+
+	request := &apiRequest{
+		fn:     a.ProtectAuthenticate,
+		method: http.MethodPost,
+		path:   "/v2/foo",
+	}
+	actual := &admin.Reply{}
+	res := s.doRequest(request, actual)
+	require.Equal(http.StatusOK, res.StatusCode)
+	expected := &admin.Reply{Success: true}
+	require.Equal(expected, actual)
+}
+
+// Test the Authenticate endpoint.
+func (s *gdsTestSuite) TestAuthenticate() {
+	require := s.Require()
+	a := initAdmin(s)
+
+	// Missing credential
+	request := &apiRequest{
+		fn:     a.Authenticate,
+		method: http.MethodPost,
+		path:   "/v2/authenticate",
+		in:     &admin.AuthRequest{},
+	}
+	res := s.doRequest(request, nil)
+	require.Equal(http.StatusUnauthorized, res.StatusCode)
+
+	// Invalid credential
+	request.in = &admin.AuthRequest{
+		Credential: "invalid",
+	}
+	res = s.doRequest(request, nil)
+	require.Equal(http.StatusUnauthorized, res.StatusCode)
+
+	// MAYBE: Can we mock the token validation to test the success path?
+}
+
+// Test the Reauthenticate endpoint.
+func (s *gdsTestSuite) TestReauthenticate() {
+	require := s.Require()
+	a := initAdmin(s)
+
+	claims := &tokens.Claims{
+		StandardClaims: jwt.StandardClaims{
+			Id:        uuid.NewString(),
+			Audience:  "http://localhost",
+			IssuedAt:  time.Now().Unix(),
+			NotBefore: time.Now().Unix(),
+			ExpiresAt: time.Now().Unix(),
+		},
+	}
+	accessToken := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+	refreshToken, err := a.tokens.CreateRefreshToken(accessToken)
+	require.NoError(err)
+	access, err := a.tokens.Sign(accessToken)
+	require.NoError(err)
+	refresh, err := a.tokens.Sign(refreshToken)
+	require.NoError(err)
```

Comment:
It might be helpful to create a method to create credentials on demand, caching them for reuse since we'll need it for many tests. 

```go
func (s *gdsTestSuite) credentials(refresh bool) (ats, rts string, err error) {
	if refresh {
		// Delete cached access/refresh tokens 
		s.accessToken = ""
		s.refreshToken = ""
	}

	if s.accessToken == "" || s.refreshToken == "" {
		// Regenerate credentials
		creds := map[string]interface{}{
			"sub":     "102374163855881761273",
			"hd":      "rotational.io",
			"email":   "kate@rotational.io",
			"name":    "Kate Holland",
			"picture": "https://foo.googleusercontent.com/test!/Aoh14gJceTrUA",
		}
		accessToken, err := s.admin.tokens.CreateAccessToken(creds)
		if err != nil {
				return "", "", err
		}

		refreshToken, err := s.admin.tokens.CreateRefreshToken(accessToken)
		…

		s.accessToken = s.admin.tokens.Sign(accessToken)
		s.refreshToken = s.admin.tokens.Sign(refreshToken)
	}

	return s.accessToken, s.refreshToken, nil
}
```